### PR TITLE
improve usage doc for starlark

### PR DESCRIPTION
**Pull request by mengqiy**
There are 3 sections in the usage doc now:
- Overview
- Function
- FunctionConfig

I will update our template if this makes sense.

### CODE REVIEW

File: functions/go/starlark/README.md [hunk 1]
```diff
@@ -8,34 +8,46 @@ Run a Starlark script to mutate or validate resources.
...
+reference [here][apiref].
 
 <!--mdtogo:Long-->
 
-[Starlark] is a dialect of Python. It is commonly used as a configuration language. It is an untyped dynamic language with
-high-level data types, first-class functions with lexical scope, and garbage
-collection.
+### Function
+
+[starlark] function contains an interpreter for the Starlark language. You can
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:21]: you already linked to Starlark docs in overview.

also starlark -> `starlark`

File: functions/go/starlark/README.md [hunk 2]
```diff
@@ -8,34 +8,46 @@ Run a Starlark script to mutate or validate resources.
...
+provide a Starlark script in the `functionConfig` to mutate or validate your
+resources. It is a light-weight way to write a KRM function. It is recommended
+to only build simple function with it. Generally, if you Starlark script is
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:24]: Just link to the book: https://kpt.dev/book/05-developing-functions/?id=approaches

This is an example where you don't want to repeat the book. If we have a python SDK in the future, we shouldn't need to change this doc.

File: functions/go/starlark/README.md [hunk 3]
```diff
@@ -8,34 +8,46 @@ Run a Starlark script to mutate or validate resources.
...
+first-class functions with lexical scope, and garbage collection. You can find
+the spec for the Starlark language [here][spec]. You can also find its API
+reference [here][apiref].
 
 <!--mdtogo:Long-->
 
-[Starlark] is a dialect of Python. It is commonly used as a configuration language. It is an untyped dynamic language with
-high-level data types, first-class functions with lexical scope, and garbage
-collection.
+### Function
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:19]: This header is extraneous.

File: functions/go/starlark/README.md (hunk 1 above)
Comment (frankfarzan) [inline functions/go/starlark/README.md:21]: The `starlark` function

File: functions/go/starlark/README.md [hunk 4]
```diff
@@ -8,34 +8,46 @@ Run a Starlark script to mutate or validate resources.
...
+resources. It is a light-weight way to write a KRM function. It is recommended
+to only build simple function with it. Generally, if you Starlark script is
+longer than 20 lines, you may want to consider build a function with our
+[Golang SDK] or [TypeScript SDK].
+
+### FunctionConfig
 
-The starlark function currently uses `StarlarkRun` as the function config. It
-looks like this:
+The starlark function accepts a CRD of kind `StarlarkRun` as the
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:30]: Technically a CR not a CRD. But spell it out "custom resource"

File: functions/go/starlark/README.md [hunk 5]
```diff
@@ -8,34 +8,46 @@ Run a Starlark script to mutate or validate resources.
...
   name: my-star-fn
 source: |
-  # set the namespace on each resource
-  def run(resources, ns_value):
-    for resource in resources:
-    # mutate the resource
-    resource["metadata"]["namespace"] = ns_value
-  run(ctx.resource_list["items"], "prod")
+  # Starlark source lives here.
+  # You can mutate or validate the resources in ctx.resource_list["items"].
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:40]: can be a follow up. How about `results`?

**Review by frankfarzan — COMMENTED** [review]
Much better

File: functions/go/starlark/README.md (hunk 5 above)
Comment (frankfarzan) [inline functions/go/starlark/README.md:40]: There's lot more to know here about how to write such a function. For example:

- you need a function to call
- you need to understand the the Functions Spec: https://kpt.dev/book/05-developing-functions/01-functions-specification

Comment (mengqiy) [reply functions/go/starlark/README.md]: To be clear, old kpt starlark doesn't support this either. 
Created https://github.com/GoogleContainerTools/kpt/issues/2201 to track it.

**Comment by mengqiy** [comment]
@frankfarzan PTAL

File: functions/go/starlark/README.md [hunk 6]
```diff
@@ -4,27 +4,36 @@
 
 <!--mdtogo:Short-->
 
-Run a Starlark script to mutate or validate resources.
+The `starlark` function contains a Starlark interpreter to run a Starlark script
+to mutate or validate resources.
+
+It follows the [executable configuration] pattern. It makes writing simple
+function trivial.
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:11]: I'd avoid general statements like making things "trivial". There are friction points in executable config flow.

File: functions/go/starlark/README.md [hunk 7]
```diff
@@ -33,9 +42,69 @@ source: |
...
+  the [KRM Functions Specification]. You can read the input resources from
+  `ctx.resource_list[items]` and the `functionConfig` from
+  `ctx.resource_list[functionConfig]`.
+- Write resources to `ctx.resource_list[items]`.
+- Return an error using [`fail`][fail].
+- Write error message to stderr using [`print`][print]
+
+Here's what you cannot do in the Starlark script:
+
+- We don't support the `results` field, i.e. if you write some results in
```
Comment (frankfarzan) [inline functions/go/starlark/README.md:68]: We should support it at some point? If so, it's a current limitation and you want to use "yet" here.

Comment (mengqiy) [reply functions/go/starlark/README.md]: Done

File: functions/go/starlark/README.md (hunk 6 above)
Comment (mengqiy) [reply functions/go/starlark/README.md]: I updated the wording here.

> There are friction points in executable config flow.

Is there anything we can address in this PR?

**Review by frankfarzan — APPROVED** [review]

**Review by droot — APPROVED** [review]

**Pull request merged**